aiida_siesta/workflows/iterate.py

```diff
@@ -0,0 +1,805 @@
+        should be set.
+
+        E.g.:
+            if the parameter 'kpoints_whatever' means that you need to parse
+            the value into a KpointsData() and set it to the 'kpoints' input,
+            this should return 'kpoints' 
+
+        May be overwritten by child classes.
+
+        Parameters
+        ------------
+        key: str
+            the name of the parameter.
+        """
+        return key
+
+
+class SiestaIterator(InputIterator):
+    """
+    General iterator for ANY parameter in SIESTA.
+
+    It runs `SiestaBaseWorkChain` iteratively setting the inputs 
+    appropiately at each iteration.
+
+    With the help of the global dictionaries `_BASIS_PARAMS`, 
+    `_PARAMS` and `_INPUT_ATTRIBUTES`, it decides what each parameter
+    key in the "iterate_over" needs to do (i.e. which input to modify and
+    how to parse the value into a valid input).
+    """
+
+    _process_class = SiestaBaseWorkChain
+    _expose_inputs_kwargs = {'exclude': ('metadata',)}
+
+    @classmethod
+    def define(cls, spec):
+        super().define(spec)
+
+        spec.inputs._ports['pseudos'].dynamic = True
+        # spec.input('units', valid_type=Str, required=False, help='The units of the parameter')
+
+    def initialize(self):
+        """
+        Initialize the dicts to store input keys and parsing functions for
+        each parameter.
+        """
+        self.ctx._input_keys = {}
+        self.ctx._parsing_funcs = {}
+
+        super().initialize()
+
+    def _parse_key(self, key):
+        """
+        Uses the _parse_key method to store the input keys and parsing functions
+        that we are going to use through the workchain.
+
+        Parameters
+        -----------
+        key: str
+            the name of the parameter, as passed in the "iterate_over" input.
+        """
+        
+        input_key, parsing_func = input_key_and_parsing_func(key)
+
+        self.ctx._input_keys[key] = input_key
+        self.ctx._parsing_funcs[key] = parsing_func
+
+        return key
+    
+    def _attr_from_key(self, key):
+        '''
+        Returns the input key to which the parameter belongs.
+
+        Parameters
+        ------------
+        key: str
+            the name of the parameter.
+        '''
+        return self.ctx._input_keys[key]
+
+    def _parse_val(self, val, key, inputs):
+        '''
+        Parses the value using the parsing function that we have stored for this parameter.
+
+        Parameters
+        -----------
+        val: any
+            the value to parse
+        key: str
+            the parameter to which this value belongs
+        inputs: AttributeDict
+            all the already existing inputs.
+        '''
+
+        parsing_func = self.ctx._parsing_funcs[key]
+
+        if parsing_func is not None:
+            val = parsing_func(val, inputs)
+
+        return val
+
+
+def input_key_and_parsing_func(parameter):
```

Comment:
Can you check the new version of this and see if you like it? The main thing that I didn't like about your proposal is that it is not easy to extend, because it's not general. I'd like to keep the possible parameters as an open list so that it can be extended at any moment. And I also like the fact that they are well organized in a list that you can check (and maybe use it to autodocument the workflow.)

Also, right now it's so general that it works for any `InputIterator`, not only the `SiestaIterator`. That's why the `input_key_and_parse_func` function is now a method of `InputIterator`. All `SiestaIterator` adds is a lookup table that indicates what to do with each parameter (`SIESTA_ITERATION_PARAMS`).